File: Napakalaki/src/napakalaki/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author fran
 */
public class SpecificBadConsequence extends BadConsequence{
    
    // Variables
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    // Constructor
    public SpecificBadConsequence(String text, int levels, 
            ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden){
        
        super(text, levels);
        setSpecificVisibleTreasures(tVisible);
        setSpecificHiddenTreasures(tHidden);
        
    }
    
    // Getters
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return specificVisibleTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures(){
        return specificHiddenTreasures;
    }
    
    // Setters
    public void setSpecificVisibleTreasures(ArrayList<TreasureKind> sVisible){
        this.specificVisibleTreasures = new ArrayList(sVisible);
    }
    
    public void setSpecificHiddenTreasures(ArrayList<TreasureKind> sHidden){
        this.specificHiddenTreasures = new ArrayList(sHidden);
    }
    
    // Utils
    @Override
    public boolean isEmpty(){
       
        return specificVisibleTreasures.isEmpty() && 
                specificHiddenTreasures.isEmpty();
        
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        
        // Si el elemento no existe, no hace nada
        specificVisibleTreasures.remove(t.getType());
        
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        
        // Si el elemento no existe, no hace nada
        specificHiddenTreasures.remove(t.getType());
        
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(
        ArrayList<Treasure> visibleTreasures, 
        ArrayList<Treasure> hiddenTreasures){
        
        SpecificBadConsequence bcAux = new SpecificBadConsequence(super.getText(), 
                0, new ArrayList<>(), new ArrayList<>());
        
        // Si los tesoros visibles que se reciben no están vacíos
        if(!visibleTreasures.isEmpty()){
                                      
            // Si el jugador tiene tesoros visibles equipados
            if(!specificVisibleTreasures.isEmpty()){
                
                ArrayList<TreasureKind> trAux = new ArrayList();
                
                for(Treasure treasure : visibleTreasures){
                    
                    if(specificVisibleTreasures.contains(treasure.getType())){
                        trAux.add(treasure.getType());
                    }
                    
                }
                
                bcAux.setSpecificVisibleTreasures(trAux);
                
            }
            
        }
        
        // Si los tesoros ocultos que se reciben no están vacíos
        if(!hiddenTreasures.isEmpty()){

            // Si el jugador tiene tesoros ocultos equipados
            if(!specificHiddenTreasures.isEmpty()){
                
                ArrayList<TreasureKind> trAux = new ArrayList();
                
                for(Treasure treasure : visibleTreasures){
                    
                    if(specificHiddenTreasures.contains(treasure.getType())){
                        trAux.add(treasure.getType());
                    }
                    
                }
                
                bcAux.setSpecificHiddenTreasures(trAux);
                
            }
            
        }
        
        return bcAux;
        
    }
    
    @Override
    public String toString(){
        
        return super.toString() +
            ", SpecificVisibleTreasures = " + specificVisibleTreasures.toString() +
            ", SpecificHiddenTreasures = " + specificHiddenTreasures.toString();
        
    }
    
}

File: NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "bad_consequence.rb"

class SpecificBadConsequence < BadConsequence
  
  public
  
  def initialize(t, l, v ,h)
    super(t, l)
    @specificVisibleTreasures = v
    @specificHiddenTreasures = h
  end
    
  public_class_method :new
  
  # Getters
  def get_specific_visible_treasures
    @specificVisibleTreasures
  end
  
  def get_specific_hidden_treasures
    @specificHiddenTreasures
  end
  
  # Setters
  def set_specific_visible_treasures(v)
    @specificVisibleTreasures = v
  end
  
  def set_specific_hidden_treasures(h)
    @specificHiddenTreasures = h
  end
  
  # Utils
  def is_empty
    
    if @specificHiddenTreasures.nil?
      @specificHiddenTreasures = Array.new
    end
    
    @specificVisibleTreasures.empty? and @specificHiddenTreasures.empty?
          
  end
  
  def substract_visible_treasure(t)
    @specificVisibleTreasures.delete_at(@specificVisibleTreasures.index(t.get_type) || @specificVisibleTreasures.length)
  end
  
  def substract_hidden_treasure(t)
    @specificHiddenTreasures.delete_at(@specificHiddenTreasures.index(t.get_type) || @specificHiddenTreasures.length)
  end
  
  def adjust_to_fit_treasure_lists(v, h)
    
    bc_aux = SpecificBadConsequence.new("", 0, Array.new, Array.new)
    
    # Si los tesoros visibles que se reciben no están vacíos
    if v.any? && @specificVisibleTreasures.any?
                    
          tr_aux = Array.new
          
          v.each { |t|
          
            if @specificVisibleTreasures.include?(t.get_type)
              tr_aux << t.get_type
            end
            
          }
          
          bc_aux.set_specific_visible_treasures(tr_aux)
                  
    end
    
    # Si los tesoros ocultos que se reciben no están vacíos
    if h.any? && @specificHiddenTreasures.any?
                       
          tr_aux_h = Array.new
          
          h.each { |t|
          
            if @specificHiddenTreasures.include?(t.get_type)
              tr_aux_h << t.get_type
            end
            
          }
          
          bc_aux.set_specific_hidden_treasures(tr_aux_h)
                     
    end
    
    bc_aux
    
  end
  
  def to_s
    super + ", SpecificHiddenTreasures: #{@specificHiddenTreasures}," +
        " SpecificVisibleTreasures: #{@specificVisibleTreasures}"
  end
  
end
